server/src/main/scala/org/http4s/server/package.scala

```diff
@@ -140,25 +140,42 @@ package object server {
     inDefaultServiceErrorHandler[F, F]
 
   def inDefaultServiceErrorHandler[F[_], G[_]](implicit
-      F: Monad[F]): Request[G] => PartialFunction[Throwable, F[Response[G]]] =
+      F: Monad[F]): Request[G] => PartialFunction[Throwable, F[Response[G]]] = {
+    def messageFailureLogAction: (Throwable, => String) => F[Unit] = { (t, msg) =>
+      Monad[F].pure(messageFailureLogger.debug(t)(msg))
+    }
+
+    def serviceErrorLogAction: (Throwable, => String) => F[Unit] = { (t, msg) =>
+      Monad[F].pure(serviceErrorLogger.error(t)(msg))
+    }
+
+    serviceErrorHandler[F, G](messageFailureLogAction, serviceErrorLogAction)
+  }
+
+  def serviceErrorHandler[F[_], G[_]](
+      messageFailureLogAction: (Throwable, => String) => F[Unit],
+      serviceErrorLogAction: (Throwable, => String) => F[Unit]
+  )(implicit F: Monad[F]): Request[G] => PartialFunction[Throwable, F[Response[G]]] =
     req => {
       case mf: MessageFailure =>
-        messageFailureLogger.debug(mf)(
+        val msg: String =
```

Comment:
it seems that `msg` is evaluated eagerly here even if the following call to `messageFailureLogAction` takes no action with it. The same applies  for the `msn` below.

Reply:
I mean, if you're going to call this new `serviceErrorHandler` function directly with just a stub callback, keep in mind that these `msg` vals are going to be evaluated anyway and then just discarded.

Reply:
> Is it your intention for some reason?

No - thanks for finding my mistake. Perhaps def would work better for the messages then?

Reply:
In terms of lazy evaluations – yes, it should. But in terms of overall performance it may work still worse than the original code, I guess. Because originally the code leveraged macros to optimize evaluation of these message strings. In the new code there's an additional layer of indirection caused by these "by-name" parameters. I'm not sure if it is a really big deal there, though.